Ucr geom list with aes edit

FrancoisJRenaud opened the pull request:
Rewrote the SetParameters sub + a few other changes + questions for a few more changes.

dannyparsons commented on instat/UcrGeomListWithAes.vb:
```diff
@@ -130,97 +137,36 @@ Public Class UcrGeomListWithParameters
+                For i = iMaxIndex To lstAesParameterLabels.Count
+                    lstAesParameterLabels(i).Visible = False
+                    lstAesParameterUcr(i).Visible = False
+                Next
```
This should only go up to `lstAesParameterLabels.Count - 1`. If the count is 10 then the last index is `(9)`.

dannyparsons commented on instat/UcrGeomListWithAes.vb:
```diff
@@ -60,6 +65,8 @@ Public Class UcrGeomListWithParameters
         ucrReceiverParam9.Selector = UcrSelector
         ucrReceiverParam10.Selector = UcrSelector
+        'Set the default receiver.
+        'Question: should that not go into UcrGeomListWithParameters_Load ?
```
I think there is another part of the code that sets the the current receiver as the first empty receiver after filling in from the geom. So I would guess this isn't needed at all. Have a look in the rest of the code.

FrancoisJRenaud replied:
Yup it's done in the set aes sub
Ready to merge now

FrancoisJRenaud commented:
Continued editing after ready to merge but focused on stuff in PLots and GeomListWithAes.

dannyparsons commented on instat/UcrGeomListWithAes.vb:
```diff
@@ -97,15 +96,17 @@ Public Class UcrGeomListWithParameters
-                    If (clsCurrGeom.strGeomName = "geom_boxplot" OrElse clsCurrGeom.strGeomName = "geom_dotplot" OrElse clsCurrGeom.strGeomName = "geom_bar") AndAlso clsParam.strArgumentName = "x" AndAlso clsParam.strArgumentValue = Chr(34) & Chr(34) Then
+                    If clsParam.strArgumentName = "x" AndAlso clsParam.strArgumentValue = Chr(34) & Chr(34) Then
+                        'For some geoms like BoxPlot, when the x aes is not filled, ggplot R syntax requires to set x="". This x="" might be copied into the global aes if the ApplyOnAllLayers is set to true for a BoxPlot Layer (might want to avoid this to be written in the global aes at first... This should only ever appear in the appropriate geoms...). This might be copied from the GgplotAesFunc ...
+                        'Question Task Warning: maybe add a warning for the user to not call his variables (colums) "" !
```
Good thing to consider, but "" is a invalid column name in R Instat, so that should never be the case.

dannyparsons commented on instat/UcrGeomListWithAes.vb:
```diff
@@ -97,15 +96,17 @@ Public Class UcrGeomListWithParameters
-                    'Should do this for any geom where x = "" ?
-                    If (clsCurrGeom.strGeomName = "geom_boxplot" OrElse clsCurrGeom.strGeomName = "geom_dotplot" OrElse clsCurrGeom.strGeomName = "geom_bar") AndAlso clsParam.strArgumentName = "x" AndAlso clsParam.strArgumentValue = Chr(34) & Chr(34) Then
+                    If clsParam.strArgumentName = "x" AndAlso clsParam.strArgumentValue = Chr(34) & Chr(34) Then
+                        'For some geoms like BoxPlot, when the x aes is not filled, ggplot R syntax requires to set x="". This x="" might be copied into the global aes if the ApplyOnAllLayers is set to true for a BoxPlot Layer (might want to avoid this to be written in the global aes at first... This should only ever appear in the appropriate geoms...). This might be copied from the GgplotAesFunc ...
```
I don't see why `x = ""` should never go in global aes. Try these two examples:
`ggplot(survey, mapping = aes(x = "", y = Yield)) + geom_boxplot() + geom_point()`
and
`ggplot(survey, mapping = aes(y = Yield)) + geom_boxplot(mapping = aes(x = "")) + geom_point()`
The second gives an error, we need `x = ""` in the global aes for it to work.

FrancoisJRenaud replied:
I see, my bad

dannyparsons commented on instat/UcrGeomListWithAes.vb:
```diff
@@ -117,7 +118,9 @@ Public Class UcrGeomListWithParameters
-                    If clsCurrGeom.strGeomName = "geom_boxplot" AndAlso clsParam.strArgumentName = "x" AndAlso clsParam.strArgumentValue = Chr(34) & Chr(34) Then
+                    'Should do this for any geom with x="" ?
+                    If (clsCurrGeom.strGeomName = "geom_boxplot" OrElse clsCurrGeom.strGeomName = "geom_dotplot" OrElse clsCurrGeom.strGeomName = "geom_bar") AndAlso clsParam.strArgumentName = "x" AndAlso clsParam.strArgumentValue = Chr(34) & Chr(34) Then
+                        'Similar check to the one just above. However, in this case, we can filter by looking at who the current geom is.
```
I think here we also don't need to check the current geom name. Even though only certain geoms might have this, removing the check means we don't have to know all the geoms which may contain `x = ""`. And if any geom has `x = ""` I think we still want to this.

FrancoisJRenaud replied:
Okéé, agreed

dannyparsons commented on instat/sdgPlots.vb:
```diff
@@ -38,98 +46,76 @@ Public Class sdgPlots
-        IncludeFacets()
+
         chkIncludeFacets.Checked = False
+        'Changing the chkIncludeFacets will trigger IncludeFacets() ...
```
This only triggers the CheckedChanged event if the control was checked before, so I think that lines needs to stay. Although it may work at the moment, if the check is changed by some other part of the code outside the subdialog, this could give errors.

FrancoisJRenaud replied:
Ok, yeah, I get confused between what triggers events or not.

dannyparsons commented on instat/sdgPlots.vb:
```diff
@@ -38,98 +46,76 @@ Public Class sdgPlots
-    End Sub
-
     Private Sub Facets()
+        'Links the factor receivers, used for creating facets, with the selector. The variables need to be factors.
```
I would prefer a more specific name for this function than "Facets", maybe "FacetReceiverSetup"

dannyparsons commented on instat/sdgPlots.vb:
```diff
@@ -166,16 +152,15 @@ Public Class sdgPlots
             chkNoOfRowsOrColumns.Visible = False
             nudNoOfRowsOrColumns.Visible = False
         End If
+        'Question to be discussed: should SetFacets not only be run when chkIncludeFacets is checked ?
```
It depends how it's implemented. If SetFacets checks if chkIncludeFacets is checked and removes facets parameter when it's not, then it can be called in either case.

dannyparsons commented on instat/sdgPlots.vb:
```diff
@@ -243,14 +271,12 @@ Public Class sdgPlots
     Private Sub ucr1stFactorReceiver_SelectionChanged(sender As Object, e As EventArgs) Handles ucr1stFactorReceiver.SelectionChanged
         SetFacets()
         IncludeFacetsOperator()
+        'Question: Do we not only need IncludeFacetsOperator here and not in the following subs, as only the first receiver and the chkIncludeFacets decides whether the operator should be added ? Can I delete the others ?
```
Feel free to tidy up the facet methods

FrancoisJRenaud commented on instat/UcrGeomListWithAes.vb:
```diff
@@ -20,7 +20,9 @@ Imports instat
     Public lstAesParameterUcr As New List(Of ucrReceiverSingle)
+    'The two previous fields are the lists of parameter labels and receivers on the ucr.
     Public lstCurrArguments As New List(Of String)
+    'Question: lstCurrArguments is the list of aes parameters names of the current geom (ucrGeom.clsCurrGeom), i.e. the list of parameter names of clsGeomAesFunction. Right ?
```
Sure, it was just the last thing done, and quite late. Will go over it again to confirm, but should be ok

FrancoisJRenaud commented on instat/sdgPlots.vb:
```diff
@@ -38,98 +46,76 @@ Public Class sdgPlots
-    Public Sub SetGgplotFunction(clsGgplotFunc As RFunction)
-        clsRggplotFunction = clsGgplotFunc
+    Private Sub InitialiseTabs()
+        'Question: Tabs are set in turn as selected, using their index. Then the first tab is set as selected again. For some reason this initialises the tabs ?
```
So there are internal initialising subs that handle the load event that need to be triggered by showing the tab ?  Maybe to discuss on skype...

FrancoisJRenaud commented:
I ve been going through the SetFacets and SetScaleOptions, and I have a few concerns.

- The choice of using grid or wrap is not left to the user. I can imagine this is a choice made by the team, but I would like to hear what the arguments are. 
- We excluded facet_grid(a+b ~.) and facet_grid(. ~ a+b). I guess these are not common enough ?
- The users doesn't have access to the space = "free"/"fixed" parameter in the facet_grid case....

Otherwise I ve just added that if the user only filled in the second receiver, the faceting still happens. Don't know if you are ok with that. Just started doing it by error, then thought it might actually be sensible to do so. But can remove if you are against it. Just find it more comfortable for the user.

dannyparsons commented:
I'm happy you are looking into facets and seeing if there's any improvements. David and Beth mostly worked on this so he would be better to discuss it with.

If you are going to make changes to facets, I would prefer that is done on a separate pull request. There is already lots of good discussions on this pull request, but it gets complicated when there are so many topics in a single pull request, even if it's the same files. I would suggest focusing on the original issues, getting them working and merged, before starting on a different topic.

volloholic commented:
I'm happy to discuss this on Skype when you want with a few other things. I can fill you in on our thinking.

Maybe the key point for those discussions is that we want users to be able to achieve what they want without necessarily learning about how GGplot works.

FrancoisJRenaud commented:
I ve saved the other changes about issue  #1948 and not included in here as required. Will stop stacking projects in the same pull request. Otherwise in facets, just added the Free Space chk box and made the 2nd factor receiver enabling dependent on the emptiness of the first factor receiver. Made a mistake when discussing with @volloholic: the default was to have fixed number of rows/columns ticked. So I changed it as we assumed it was the opposite. Just mentioning here in case you have strong feelings about it.
